### PR TITLE
[docs] Limit preval usage and simplify caching

### PR DESCRIPTION
**Pull request by byCedric**
# Why

This PR does 3 things to help with the usage and performance of the docs. Right now, we use two different caches in CI to separate MDX content and Next + Preval content. With this change, we separate the caches with MDX + Next content and Preval content.

1. Preval has [issues invalidating the cache](https://github.com/kentcdodds/babel-plugin-preval/issues/19). Using this in more places than absolute necessary might mess up the cache more. This change removes the "global" Preval plugin. Instead, it adds a special loader for all `docs/constants/**` files and includes Preval within those files.

2. Because we consolidated Next and MDX caches, we can simplify the CI. With this change we only have a single Next cache, and never store prevaled files.

3. We also had an issue where a single `yarn export` run created two different build ids, causing some issues with the JS/HTML again. It could be related to us caching the full `.next` folder, [instead of just `.next/cache`](https://github.com/vercel/next.js/blob/canary/errors/no-cache.md#possible-ways-to-fix-it). This changes to only `.next/cache`, removing any traces of previous builds.

I couldn't split this into 3 different PRs unfortunately, they are pretty much all related.

# How

- Moved the constant files in `common/` to `constants/` and replaced all occurrences
- Limited Preval usage to only `constants/` files by adding a custom webpack loader in `next.config.js`
- [Traced the last "JS not found" error to multiple build IDs being published](https://github.com/expo/expo/runs/1063326056?check_suite_focus=true#step:12:112)

# Test Plan

**To test the Preval cache**
- `$ yarn next dev` to build first cache
- Remove random pages from `pages/versions/v38.0.0/sdk`
- Validate there are no changes in the sidebar (on any `v38.0.0` page), resulting in `404` from the removed pages
- Manually remove `.next/preval` folder, and restart `$ yarn next dev` (will remove all other files/folders except `cache` in `.next`)
- Validate there are changes in the sidebar, all removed pages should be gone

**To test the cache**
- Validate if CI only caches `.next/cache`
- There should be no double-id-export occurrences
- Check if there is no warning about ["no build cache"](https://github.com/vercel/next.js/blob/canary/errors/no-cache.md)

### CODE REVIEW

**Review by brentvatne — APPROVED** [review]
awesome work! this makes our caching setup a lot easier to grok :+1:

File: docs/next.config.js [hunk 1]
```diff
@@ -17,21 +17,31 @@ module.exports = withCSS({
...
+    // It's to prevent over-usage and separate the cache to allow manually invalidation.
+    // See: https://github.com/kentcdodds/babel-plugin-preval/issues/19
+    config.module.rules.push({
+      test: /.jsx?$/,
+      include: [resolve(__dirname, 'constants')],
+      use: {
+        ...options.defaultLoaders.babel,
+        options: {
+          ...options.defaultLoaders.babel.options,
+          cacheDirectory: '.next/preval',
```
Comment (ide) [inline docs/next.config.js:30]:
```suggestion
          // Keep this path in sync with package.json and other scripts that clear the cache
          cacheDirectory: '.next/preval',
```

File: docs/next.config.js [hunk 2]
```diff
@@ -17,21 +17,31 @@ module.exports = withCSS({
...
-      ...options.defaultLoaders.babel,
-      options: {
-        ...options.defaultLoaders.babel.options,
-        cacheDirectory: 'node_modules/.cache/babel-mdx-loader',
+    // Add preval support for `constants/*` only and move it to the `.next/preval` cache.
+    // It's to prevent over-usage and separate the cache to allow manually invalidation.
+    // See: https://github.com/kentcdodds/babel-plugin-preval/issues/19
+    config.module.rules.push({
+      test: /.jsx?$/,
+      include: [resolve(__dirname, 'constants')],
```
Comment (ide) [inline docs/next.config.js:25]: `join` should be sufficient here - `__dirname` is always resolved (I think? maybe some edge cases with symlinks)

File: docs/next.config.js [hunk 3]
```diff
@@ -17,21 +17,31 @@ module.exports = withCSS({
...
+      use: {
+        ...options.defaultLoaders.babel,
+        options: {
+          ...options.defaultLoaders.babel.options,
+          cacheDirectory: '.next/preval',
+          plugins: ['preval'],
```
Comment (ide) [inline docs/next.config.js:31]: needed?
```suggestion
          plugins: [...options.defaultLoaders.babel.options?.plugins ?? [], 'preval'],
```

**Review by ide — APPROVED** [review]

**Comment by ide** [comment]
Want to make software better? Delete a cache!

File: docs/next.config.js (hunk 3 above)
Comment (byCedric) [reply docs/next.config.js:31]: Good point, I don't think we use any special plugins here. It's not rendered by Next or has functionality other than defining variables. But I agree, it's good dev hygiene 😄

**Pull request merged**